# Problem Statement

## **Business Context**

"Visit with Us," a leading travel company, is revolutionizing the tourism industry by leveraging data-driven strategies to optimize operations and customer engagement. While introducing a new package offering, such as the Wellness Tourism Package, the company faces challenges in targeting the right customers efficiently. The manual approach to identifying potential customers is inconsistent, time-consuming, and prone to errors, leading to missed opportunities and suboptimal campaign performance.

To address these issues, the company aims to implement a scalable and automated system that integrates customer data, predicts potential buyers, and enhances decision-making for marketing strategies. By utilizing an MLOps pipeline, the company seeks to achieve seamless integration of data preprocessing, model development, deployment, and CI/CD practices for continuous improvement. This system will ensure efficient targeting of customers, timely updates to the predictive model, and adaptation to evolving customer behaviors, ultimately driving growth and customer satisfaction.


## **Objective**

As an MLOps Engineer at "Visit with Us," your responsibility is to design and deploy an MLOps pipeline on GitHub to automate the end-to-end workflow for predicting customer purchases. The primary objective is to build a model that predicts whether a customer will purchase the newly introduced Wellness Tourism Package before contacting them. The pipeline will include data cleaning, preprocessing, transformation, model building, training, evaluation, and deployment, ensuring consistent performance and scalability. By leveraging GitHub Actions for CI/CD integration, the system will enable automated updates, streamline model deployment, and improve operational efficiency. This robust predictive solution will empower policymakers to make data-driven decisions, enhance marketing strategies, and effectively target potential customers, thereby driving customer acquisition and business growth.

## **Data Description**

The dataset contains customer and interaction data that serve as key attributes for predicting the likelihood of purchasing the Wellness Tourism Package. The detailed attributes are:

**Customer Details**
- **CustomerID:** Unique identifier for each customer.
- **ProdTaken:** Target variable indicating whether the customer has purchased a package (0: No, 1: Yes).
- **Age:** Age of the customer.
- **TypeofContact:** The method by which the customer was contacted (Company Invited or Self Inquiry).
- **CityTier:** The city category based on development, population, and living standards (Tier 1 > Tier 2 > Tier 3).
- **Occupation:** Customer's occupation (e.g., Salaried, Freelancer).
- **Gender:** Gender of the customer (Male, Female).
- **NumberOfPersonVisiting:** Total number of people accompanying the customer on the trip.
- **PreferredPropertyStar:** Preferred hotel rating by the customer.
- **MaritalStatus:** Marital status of the customer (Single, Married, Divorced).
- **NumberOfTrips:** Average number of trips the customer takes annually.
- **Passport:** Whether the customer holds a valid passport (0: No, 1: Yes).
- **OwnCar:** Whether the customer owns a car (0: No, 1: Yes).
- **NumberOfChildrenVisiting:** Number of children below age 5 accompanying the customer.
- **Designation:** Customer's designation in their current organization.
- **MonthlyIncome:** Gross monthly income of the customer.

**Customer Interaction Data**
- **PitchSatisfactionScore:** Score indicating the customer's satisfaction with the sales pitch.
- **ProductPitched:** The type of product pitched to the customer.
- **NumberOfFollowups:** Total number of follow-ups by the salesperson after the sales pitch.-
- **DurationOfPitch:** Duration of the sales pitch delivered to the customer.


# Model Building

In [1]:
# Create a master folder to keep all files created when executing the below code cells
import os
os.makedirs("tourism_project", exist_ok=True)

In [2]:
# Create a folder for storing the model building files
os.makedirs("tourism_project/model_building", exist_ok=True)

## Data Registration

In [3]:
os.makedirs("tourism_project/data", exist_ok=True)

Once the **data** folder created after executing the above cell, please upload the **tourism.csv** in to the folder

## Data Preparation

In [4]:
!pip install datasets -q

In [7]:
!pip install -q huggingface_hub datasets

In [8]:
#Login to Hugging Face
from huggingface_hub import login
login()

In [13]:
# Loading dataset
from datasets import load_dataset
import pandas as pd

dataset = load_dataset(
    "csv",
    data_files="https://huggingface.co/datasets/gayathri1909/tourism.csv/resolve/main/tourism.csv"
)

df = dataset["train"].to_pandas()

print(df.shape)
df.head()

tourism.csv: 0.00B [00:00, ?B/s]

Generating train split: 0 examples [00:00, ? examples/s]

(4128, 21)


,Unnamed: 0,CustomerID,ProdTaken,Age,TypeofContact,CityTier,DurationOfPitch,Occupation,Gender,NumberOfPersonVisiting,...,ProductPitched,PreferredPropertyStar,MaritalStatus,NumberOfTrips,Passport,PitchSatisfactionScore,OwnCar,NumberOfChildrenVisiting,Designation,MonthlyIncome
0,0,200000,1,41.0,Self Enquiry,3,6.0,Salaried,Female,3,...,Deluxe,3.0,Single,1.0,1,2,1,0.0,Manager,20993.0
1,1,200001,0,49.0,Company Invited,1,14.0,Salaried,Male,3,...,Deluxe,4.0,Divorced,2.0,0,3,1,2.0,Manager,20130.0
2,2,200002,1,37.0,Self Enquiry,1,8.0,Free Lancer,Male,3,...,Basic,3.0,Single,7.0,1,3,0,0.0,Executive,17090.0
3,3,200003,0,33.0,Company Invited,1,9.0,Salaried,Female,2,...,Basic,3.0,Divorced,2.0,1,5,1,1.0,Executive,17909.0
4,5,200005,0,32.0,Company Invited,1,8.0,Salaried,Male,3,...,Basic,3.0,Single,1.0,0,5,1,1.0,Executive,18068.0


### Data Storage

The processed training and testing datasets are saved locally and uploaded to the Hugging Face dataset repository. This ensures easy  automation in the MLOps pipeline.

In [14]:
# Dataset Info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4128 entries, 0 to 4127
Data columns (total 21 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Unnamed: 0                4128 non-null   int64  
 1   CustomerID                4128 non-null   int64  
 2   ProdTaken                 4128 non-null   int64  
 3   Age                       4128 non-null   float64
 4   TypeofContact             4128 non-null   object 
 5   CityTier                  4128 non-null   int64  
 6   DurationOfPitch           4128 non-null   float64
 7   Occupation                4128 non-null   object 
 8   Gender                    4128 non-null   object 
 9   NumberOfPersonVisiting    4128 non-null   int64  
 10  NumberOfFollowups         4128 non-null   float64
 11  ProductPitched            4128 non-null   object 
 12  PreferredPropertyStar     4128 non-null   float64
 13  MaritalStatus             4128 non-null   object 
 14  NumberOf

In [15]:
# Dataset Summary
df.describe()

,Unnamed: 0,CustomerID,ProdTaken,Age,CityTier,DurationOfPitch,NumberOfPersonVisiting,NumberOfFollowups,PreferredPropertyStar,NumberOfTrips,Passport,PitchSatisfactionScore,OwnCar,NumberOfChildrenVisiting,MonthlyIncome
count,4128.000000,4128.000000,4128.000000,4128.000000,4128.000000,4128.000000,4128.000000,4128.000000,4128.000000,4128.0000,4128.000000,4128.000000,4128.000000,4128.000000,4128.000000
mean,2527.763808,202527.763808,0.193072,37.231831,1.663275,15.584787,2.949370,3.741521,3.578488,3.2953,0.295300,3.060804,0.612161,1.223595,23178.464147
std,1409.439133,1409.439133,0.394757,9.174521,0.920640,8.398142,0.718818,1.006786,0.795031,1.8563,0.456233,1.363064,0.487317,0.852685,4506.614622
min,0.000000,200000.000000,0.000000,18.000000,1.000000,5.000000,1.000000,1.000000,3.000000,1.0000,0.000000,1.000000,0.000000,0.000000,1000.000000
25%,1320.750000,201320.750000,0.000000,31.000000,1.000000,9.000000,2.000000,3.000000,3.000000,2.0000,0.000000,2.000000,0.000000,1.000000,20751.000000
50%,2603.500000,202603.500000,0.000000,36.000000,1.000000,14.000000,3.000000,4.000000,3.000000,3.0000,0.000000,3.000000,1.000000,1.000000,22418.000000
75%,3748.250000,203748.250000,0.000000,43.000000,3.000000,20.000000,3.000000,4.000000,4.000000,4.0000,1.000000,4.000000,1.000000,2.000000,25301.000000
max,4887.000000,204887.000000,1.000000,61.000000,3.000000,127.000000,5.000000,6.000000,5.000000,22.0000,1.000000,5.000000,1.000000,3.000000,98678.000000


In [16]:
# Null Value Check
df.isnull().sum()

,0
Unnamed: 0,0
CustomerID,0
ProdTaken,0
Age,0
TypeofContact,0
CityTier,0
DurationOfPitch,0
Occupation,0
Gender,0
NumberOfPersonVisiting,0


In [17]:
# Dropping unwanted columns
df.drop(columns=["Unnamed: 0", "CustomerID"], inplace=True)

In [18]:
# Dropping unwanted columns
from sklearn.model_selection import train_test_split

X = df.drop("ProdTaken", axis=1)
y = df["ProdTaken"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

In [19]:
# Combine back
train_df = X_train.copy()
train_df["ProdTaken"] = y_train

test_df = X_test.copy()
test_df["ProdTaken"] = y_test

In [20]:
# Train and Test Datasets
train_df.to_csv("tourism_project/data/train.csv", index=False)
test_df.to_csv("tourism_project/data/test.csv", index=False)

In [22]:
# Uploading Train and Test Datasets to HuggingFace
from huggingface_hub import upload_file

upload_file(
    path_or_fileobj="tourism_project/data/train.csv",
    path_in_repo="train.csv",
    repo_id="gayathri1909/tourism.csv",
    repo_type="dataset"
)

upload_file(
    path_or_fileobj="tourism_project/data/test.csv",
    path_in_repo="test.csv",
    repo_id="gayathri1909/tourism.csv",
    repo_type="dataset"
)

CommitInfo(commit_url='https://huggingface.co/datasets/gayathri1909/tourism.csv/commit/3f602e92b001ab2bf0b14836ea3a34e7b46e613d', commit_message='Upload test.csv with huggingface_hub', commit_description='', oid='3f602e92b001ab2bf0b14836ea3a34e7b46e613d', pr_url=None, repo_url=RepoUrl('https://huggingface.co/datasets/gayathri1909/tourism.csv', endpoint='https://huggingface.co', repo_type='dataset', repo_id='gayathri1909/tourism.csv'), pr_revision=None, pr_num=None)

### **Observations**

- The dataset contains 4128 records and 21 features, including a mix of numerical and categorical variables representing customer demographics, travel behavior, and interaction details.
- The target variable `ProdTaken` is binary, indicating whether a customer purchased the tourism package.
- During data preparation, it was observed that the dataset does not contain any missing values, eliminating the need for imputation and simplifying preprocessing.
- The column `Unnamed: 0` was identified as an index-like column and removed, while `CustomerID` was dropped as it is a unique identifier and does not contribute to predictive modeling.
- The dataset includes both numerical features (such as Age, MonthlyIncome, and NumberOfTrips) and categorical features (such as TypeofContact, Occupation, and ProductPitched), which will require appropriate encoding during model building.
- The data was split into training and testing sets using an 80:20 ratio to ensure proper evaluation of model performance on unseen data.
- The processed datasets were saved locally and uploaded to the Hugging Face dataset repository, enabling integration into the MLOps pipeline.
- The dataset is ready for model development.

## Model Training and Registration with Experimentation Tracking

In [23]:
# Loading train and test data from Hugging Face
train_url = "https://huggingface.co/datasets/gayathri1909/tourism.csv/resolve/main/train.csv"
test_url = "https://huggingface.co/datasets/gayathri1909/tourism.csv/resolve/main/test.csv"

train_df = pd.read_csv(train_url)
test_df = pd.read_csv(test_url)

print("Train shape:", train_df.shape)
print("Test shape:", test_df.shape)
train_df.head()

Train shape: (3302, 19)
Test shape: (826, 19)


,Age,TypeofContact,CityTier,DurationOfPitch,Occupation,Gender,NumberOfPersonVisiting,NumberOfFollowups,ProductPitched,PreferredPropertyStar,MaritalStatus,NumberOfTrips,Passport,PitchSatisfactionScore,OwnCar,NumberOfChildrenVisiting,Designation,MonthlyIncome,ProdTaken
0,26.0,Company Invited,2,23.0,Salaried,Female,2,3.0,Basic,3.0,Married,1.0,1,5,0,1.0,Executive,17741.0,1
1,42.0,Self Enquiry,1,6.0,Salaried,Female,2,4.0,Deluxe,3.0,Divorced,1.0,1,3,0,1.0,Manager,21062.0,0
2,56.0,Company Invited,1,9.0,Salaried,Male,4,4.0,Standard,4.0,Married,5.0,0,1,0,2.0,Senior Manager,29654.0,0
3,27.0,Company Invited,3,36.0,Small Business,Male,4,6.0,Deluxe,5.0,Unmarried,2.0,0,3,1,2.0,Manager,23647.0,1
4,37.0,Self Enquiry,1,9.0,Salaried,Female,4,4.0,Basic,3.0,Divorced,6.0,0,5,1,1.0,Executive,21221.0,0


In [24]:
# Checking columns
print(train_df.columns.tolist())

['Age', 'TypeofContact', 'CityTier', 'DurationOfPitch', 'Occupation', 'Gender', 'NumberOfPersonVisiting', 'NumberOfFollowups', 'ProductPitched', 'PreferredPropertyStar', 'MaritalStatus', 'NumberOfTrips', 'Passport', 'PitchSatisfactionScore', 'OwnCar', 'NumberOfChildrenVisiting', 'Designation', 'MonthlyIncome', 'ProdTaken']


In [25]:
# Splitting into X and y
X_train = train_df.drop("ProdTaken", axis=1)
y_train = train_df["ProdTaken"]

X_test = test_df.drop("ProdTaken", axis=1)
y_test = test_df["ProdTaken"]

In [26]:
# Identify numeric and categorical columns
num_cols = X_train.select_dtypes(include=["int64", "float64"]).columns.tolist()
cat_cols = X_train.select_dtypes(include=["object"]).columns.tolist()

print("Numerical columns:", num_cols)
print("Categorical columns:", cat_cols)

Numerical columns: ['Age', 'CityTier', 'DurationOfPitch', 'NumberOfPersonVisiting', 'NumberOfFollowups', 'PreferredPropertyStar', 'NumberOfTrips', 'Passport', 'PitchSatisfactionScore', 'OwnCar', 'NumberOfChildrenVisiting', 'MonthlyIncome']
Categorical columns: ['TypeofContact', 'Occupation', 'Gender', 'ProductPitched', 'MaritalStatus', 'Designation']


### Preprocessing Pipeline & Model

In [27]:
# Preprocessing
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

numeric_transformer = StandardScaler()
categorical_transformer = OneHotEncoder(handle_unknown="ignore")

preprocessor = ColumnTransformer(
    transformers=[
        ("num", numeric_transformer, num_cols),
        ("cat", categorical_transformer, cat_cols)
    ]
)

In [28]:
# Random Forest Model
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

rf_pipeline = Pipeline(steps=[
    ("preprocessor", preprocessor),
    ("model", RandomForestClassifier(random_state=42))
])

rf_pipeline.fit(X_train, y_train)

y_pred_rf = rf_pipeline.predict(X_test)

print("Random Forest Accuracy:", accuracy_score(y_test, y_pred_rf))
print("\nClassification Report:\n", classification_report(y_test, y_pred_rf))

Random Forest Accuracy: 0.9346246973365617

Classification Report:
               precision    recall  f1-score   support

           0       0.93      0.99      0.96       661
           1       0.97      0.70      0.81       165

    accuracy                           0.93       826
   macro avg       0.95      0.85      0.89       826
weighted avg       0.94      0.93      0.93       826



## Baseline Model Performance (Random Forest)

- The Random Forest classifier achieved an accuracy of 93.46% on the test dataset.

- The model performs very well in predicting the majority class (non-purchasers) with high recall (99%).

- However, the recall for the minority class (purchasers) is relatively lower at 70%, indicating that some potential buyers are not being identified correctly.

- Overall, the model demonstrates strong performance and serves as a solid baseline for further experimentation and tuning.

In [29]:
# Hyperparameter tuning
from sklearn.model_selection import GridSearchCV

param_grid_rf = {
    "model__n_estimators": [100, 200],
    "model__max_depth": [None, 10, 20],
    "model__min_samples_split": [2, 5]
}

In [30]:
# Run Gridsearch
grid_rf = GridSearchCV(
    rf_pipeline,
    param_grid_rf,
    cv=3,
    scoring="accuracy",
    n_jobs=-1
)

grid_rf.fit(X_train, y_train)

GridSearchCV(cv=3,
             estimator=Pipeline(steps=[('preprocessor',
                                        ColumnTransformer(transformers=[('num',
                                                                         StandardScaler(),
                                                                         ['Age',
                                                                          'CityTier',
                                                                          'DurationOfPitch',
                                                                          'NumberOfPersonVisiting',
                                                                          'NumberOfFollowups',
                                                                          'PreferredPropertyStar',
                                                                          'NumberOfTrips',
                                                                          'Passport',
                                                                          'PitchSatisfactionScore',
                                                                          'OwnCar',
                                                                          'NumberOfChildrenVisiting',
                                                                          'MonthlyIncome']),
                                                                        ('cat',
                                                                         OneHotEncoder(handle_unknown='ignore'),
                                                                         ['TypeofContact',
                                                                          'Occupation',
                                                                          'Gender',
                                                                          'ProductPitched',
                                                                          'MaritalStatus',
                                                                          'Designation'])])),
                                       ('model',
                                        RandomForestClassifier(random_state=42))]),
             n_jobs=-1,
             param_grid={'model__max_depth': [None, 10, 20],
                         'model__min_samples_split': [2, 5],
                         'model__n_estimators': [100, 200]},
             scoring='accuracy')

In [31]:
print("Best Parameters:", grid_rf.best_params_)

Best Parameters: {'model__max_depth': 20, 'model__min_samples_split': 2, 'model__n_estimators': 200}


In [32]:
best_rf = grid_rf.best_estimator_

y_pred_rf_tuned = best_rf.predict(X_test)

from sklearn.metrics import accuracy_score

rf_tuned_acc = accuracy_score(y_test, y_pred_rf_tuned)

print("Tuned RF Accuracy:", rf_tuned_acc)

Tuned RF Accuracy: 0.9309927360774818


### Hyperparameter Tuning - Random Forest

Grid Search was used to tune the Random Forest model by optimizing parameters such as number of estimators, maximum depth, and minimum samples split.

The best parameters identified were:
- n_estimators = 200  
- max_depth = 20  
- min_samples_split = 2  

The tuned model achieved an accuracy of approximately 93.10%, which is slightly lower than the baseline model. This indicates that the baseline model was already well-optimized and that further tuning did not significantly improve performance.

In [33]:
# Gradient Boosting
from sklearn.ensemble import GradientBoostingClassifier

gb_pipeline = Pipeline(steps=[
    ("preprocessor", preprocessor),
    ("model", GradientBoostingClassifier(random_state=42))
])

gb_pipeline.fit(X_train, y_train)

y_pred_gb = gb_pipeline.predict(X_test)

gb_acc = accuracy_score(y_test, y_pred_gb)

print("Gradient Boosting Accuracy:", gb_acc)

Gradient Boosting Accuracy: 0.8753026634382567


In [34]:
# AdaBoost
from sklearn.ensemble import AdaBoostClassifier

ab_pipeline = Pipeline(steps=[
    ("preprocessor", preprocessor),
    ("model", AdaBoostClassifier(random_state=42))
])

ab_pipeline.fit(X_train, y_train)

y_pred_ab = ab_pipeline.predict(X_test)

ab_acc = accuracy_score(y_test, y_pred_ab)

print("AdaBoost Accuracy:", ab_acc)

AdaBoost Accuracy: 0.8329297820823245


In [35]:
import pandas as pd

model_comparison = pd.DataFrame({
    "Model": ["Random Forest (Baseline)", "Random Forest (Tuned)", "Gradient Boosting", "AdaBoost"],
    "Accuracy": [0.9346, rf_tuned_acc, gb_acc, ab_acc]
})

model_comparison.sort_values(by="Accuracy", ascending=False)

,Model,Accuracy
0,Random Forest (Baseline),0.934600
1,Random Forest (Tuned),0.930993
2,Gradient Boosting,0.875303
3,AdaBoost,0.832930


### Model Comparison and Final Selection

- Multiple models were trained and evaluated, including Random Forest, Gradient Boosting, and AdaBoost. Hyperparameter tuning was applied to the Random Forest model using GridSearchCV.

- The baseline Random Forest model achieved the highest accuracy of 93.46%, outperforming both the tuned Random Forest model and other ensemble methods. Although tuning was performed, it did not lead to performance improvement, indicating that the default parameters were already well-suited for this dataset.

- Gradient Boosting and AdaBoost models showed comparatively lower performance, making them less suitable for this problem.

- Therefore, the baseline Random Forest model is selected as the final model for deployment due to its superior performance and robustness.

## Best Model to Hugging Face

In [36]:
#Save the best model locally
import joblib

joblib.dump(rf_pipeline, "tourism_project/model_building/best_model.pkl")

['tourism_project/model_building/best_model.pkl']

In [37]:
#Upload model to Hugging Face
from huggingface_hub import upload_file

upload_file(
    path_or_fileobj="tourism_project/model_building/best_model.pkl",
    path_in_repo="best_model.pkl",
    repo_id="gayathri1909/tourism-model",
    repo_type="model"
)

Processing Files (0 / 0)      : |          |  0.00B /  0.00B            

New Data Upload               : |          |  0.00B /  0.00B            

  ...l_building/best_model.pkl:  19%|#9        | 1.53MB / 8.01MB            

CommitInfo(commit_url='https://huggingface.co/gayathri1909/tourism-model/commit/40593b4eee900ddb52030f05284c967a6fb3098c', commit_message='Upload best_model.pkl with huggingface_hub', commit_description='', oid='40593b4eee900ddb52030f05284c967a6fb3098c', pr_url=None, repo_url=RepoUrl('https://huggingface.co/gayathri1909/tourism-model', endpoint='https://huggingface.co', repo_type='model', repo_id='gayathri1909/tourism-model'), pr_revision=None, pr_num=None)

### Model Registration

The final selected model (Random Forest baseline) was saved using joblib and uploaded to the Hugging Face Model Hub. This allows the model to be easily accessed, version-controlled, and deployed as part of the MLOps pipeline.

# Deployment

In [38]:
# installing dependencies for deployment testing
!pip install -q huggingface_hub joblib scikit-learn pandas streamlit

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.1/9.1 MB 50.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.9/6.9 MB 66.2 MB/s eta 0:00:00


In [39]:
# downloading model from Hugging Face Model Hub
from huggingface_hub import hf_hub_download
import joblib

model_path = hf_hub_download(
    repo_id="gayathri1909/tourism-model",
    filename="best_model.pkl",
    repo_type="model"
)

loaded_model = joblib.load(model_path)
print("Model loaded successfully")

best_model.pkl:   0%|          | 0.00/8.01M [00:00<?, ?B/s]

Model loaded successfully


In [40]:
# testing with one sample row
sample_input = X_test.iloc[[0]]
sample_pred = loaded_model.predict(sample_input)

print("Sample prediction:", sample_pred)
sample_input

Sample prediction: [0]


,Age,TypeofContact,CityTier,DurationOfPitch,Occupation,Gender,NumberOfPersonVisiting,NumberOfFollowups,ProductPitched,PreferredPropertyStar,MaritalStatus,NumberOfTrips,Passport,PitchSatisfactionScore,OwnCar,NumberOfChildrenVisiting,Designation,MonthlyIncome
0,44.0,Self Enquiry,1,8.0,Salaried,Female,3,1.0,Standard,3.0,Married,2.0,1,4,1,0.0,Senior Manager,22879.0


### Deployment Preparation Observation

The final trained model was successfully downloaded from the Hugging Face Model Hub and loaded for inference. A sample prediction was generated to verify that the model is deployment-ready and can be used in an application interface such as Streamlit.

## Dockerfile

In [41]:
os.makedirs("tourism_project/deployment", exist_ok=True)

In [42]:
%%writefile tourism_project/deployment/Dockerfile
# Use a minimal base image with Python 3.9 installed
FROM python:3.9

# Set the working directory inside the container to /app
WORKDIR /app

# Copy all files from the current directory on the host to the container's /app directory
COPY . .

# Install Python dependencies listed in requirements.txt
RUN pip3 install -r requirements.txt

RUN useradd -m -u 1000 user
USER user
ENV HOME=/home/user \
	PATH=/home/user/.local/bin:$PATH

WORKDIR $HOME/app

COPY --chown=user . $HOME/app

# Define the command to run the Streamlit app on port "8501" and make it accessible externally
CMD ["streamlit", "run", "app.py", "--server.port=8501", "--server.address=0.0.0.0", "--server.enableXsrfProtection=false"]

Writing tourism_project/deployment/Dockerfile


## Streamlit App

Please ensure that the web app script is named `app.py`.

In [43]:
%%writefile tourism_project/deployment/app.py
import streamlit as st
import pandas as pd
from huggingface_hub import hf_hub_download
import joblib

# Load model from Hugging Face
model_path = hf_hub_download(
    repo_id="gayathri1909/tourism-model",
    filename="best_model.pkl"
)
model = joblib.load(model_path)

# App Title
st.title("Tourism Package Purchase Prediction")
st.write("""
This application predicts whether a customer is likely to purchase the Wellness Tourism Package.
Please enter customer details below.
""")

# =========================
# User Inputs
# =========================

Age = st.number_input("Age", min_value=18, max_value=100, value=30)
TypeofContact = st.selectbox("Type of Contact", ["Company Invited", "Self Inquiry"])
CityTier = st.selectbox("City Tier", [1, 2, 3])
DurationOfPitch = st.number_input("Duration of Pitch", min_value=0.0, value=10.0)
Occupation = st.selectbox("Occupation", ["Salaried", "Small Business", "Free Lancer"])
Gender = st.selectbox("Gender", ["Male", "Female"])
NumberOfPersonVisiting = st.number_input("Number of Persons Visiting", min_value=1, value=2)
NumberOfFollowups = st.number_input("Number of Followups", min_value=0, value=2)
ProductPitched = st.selectbox("Product Pitched", ["Basic", "Standard", "Deluxe", "Super Deluxe", "King"])
PreferredPropertyStar = st.selectbox("Preferred Property Star", [1, 2, 3, 4, 5])
MaritalStatus = st.selectbox("Marital Status", ["Single", "Married", "Divorced"])
NumberOfTrips = st.number_input("Number of Trips", min_value=0, value=2)
Passport = st.selectbox("Passport", [0, 1])
PitchSatisfactionScore = st.selectbox("Pitch Satisfaction Score", [1, 2, 3, 4, 5])
OwnCar = st.selectbox("Own Car", [0, 1])
NumberOfChildrenVisiting = st.number_input("Number of Children Visiting", min_value=0, value=0)
Designation = st.selectbox("Designation", ["Manager", "Executive", "Senior Manager", "AVP", "VP"])
MonthlyIncome = st.number_input("Monthly Income", min_value=0, value=30000)

# =========================
# Create Input DataFrame
# =========================

input_data = pd.DataFrame([{
    'Age': Age,
    'TypeofContact': TypeofContact,
    'CityTier': CityTier,
    'DurationOfPitch': DurationOfPitch,
    'Occupation': Occupation,
    'Gender': Gender,
    'NumberOfPersonVisiting': NumberOfPersonVisiting,
    'NumberOfFollowups': NumberOfFollowups,
    'ProductPitched': ProductPitched,
    'PreferredPropertyStar': PreferredPropertyStar,
    'MaritalStatus': MaritalStatus,
    'NumberOfTrips': NumberOfTrips,
    'Passport': Passport,
    'PitchSatisfactionScore': PitchSatisfactionScore,
    'OwnCar': OwnCar,
    'NumberOfChildrenVisiting': NumberOfChildrenVisiting,
    'Designation': Designation,
    'MonthlyIncome': MonthlyIncome
}])

# =========================
# Prediction
# =========================

if st.button("Predict"):
    prediction = model.predict(input_data)[0]

    result = "Customer WILL Purchase Package" if prediction == 1 else "Customer WILL NOT Purchase Package"

    st.subheader("Prediction Result:")
    st.success(result)

Writing tourism_project/deployment/app.py


A Streamlit web application was developed to allow users to input customer details and receive predictions on whether the customer is likely to purchase the tourism package. The application loads the trained model from the Hugging Face Model Hub and performs real-time inference.

## Dependency Handling

Please ensure that the dependency handling file is named `requirements.txt`.

In [44]:
%%writefile tourism_project/deployment/requirements.txt
pandas==2.2.2
huggingface_hub==0.20.3
streamlit==1.32.2
joblib==1.3.2
scikit-learn==1.3.2

Writing tourism_project/deployment/requirements.txt


A requirements.txt file was created to specify all necessary dependencies required for deployment. This ensures that the application environment can be reproduced consistently across different platforms.

# Hosting

In [45]:
import os
os.makedirs("tourism_project/hosting", exist_ok=True)

In [46]:
%%writefile tourism_project/hosting/hosting.py
from huggingface_hub import HfApi
import os

api = HfApi(token=os.getenv("HF_TOKEN"))

api.upload_folder(
    folder_path="tourism_project/deployment",
    repo_id="gayathri1909/tourism-package-prediction",
    repo_type="space",
    path_in_repo=""
)

Writing tourism_project/hosting/hosting.py


A hosting script was created to push all deployment files from the local deployment folder to the Hugging Face Space. This enables the Streamlit application to be deployed as a public web app for real-time predictions.

# MLOps Pipeline with Github Actions Workflow

**Note:**

1. Before running the file below, make sure to add the HF_TOKEN to your GitHub secrets to enable authentication between GitHub and Hugging Face.
2. The below code is for a sample YAML file that can be updated as required to meet the requirements of this project.

```
name: Tourism Project Pipeline

on:
  push:
    branches:
      - main  # Automatically triggers on push to the main branch

jobs:

  register-dataset:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - name: Install Dependencies
        run: <add_code_here>
      - name: Upload Dataset to Hugging Face Hub
        env:
          HF_TOKEN: ${{ secrets.HF_TOKEN }}
        run: <add_code_here>

  data-prep:
    needs: register-dataset
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - name: Install Dependencies
        run: <add_code_here>
      - name: Run Data Preparation
        env:
          HF_TOKEN: ${{ secrets.HF_TOKEN }}
        run: <add_code_here>


  model-traning:
    needs: data-prep
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - name: Install Dependencies
        run: <add_code_here>
      - name: Start MLflow Server
        run: |
          nohup mlflow ui --host 0.0.0.0 --port 5000 &  # Run MLflow UI in the background
          sleep 5  # Wait for a moment to let the server starts
      - name: Model Building
        env:
          HF_TOKEN: ${{ secrets.HF_TOKEN }}
        run: <add_code_here>


  deploy-hosting:
    runs-on: ubuntu-latest
    needs: [model-traning,data-prep,register-dataset]
    steps:
      - uses: actions/checkout@v3
      - name: Install Dependencies
        run: <add_code_here>
      - name: Push files to Frontend Hugging Face Space
        env:
          HF_TOKEN: ${{ secrets.HF_TOKEN }}
        run: <add_code_here>

```

**Note:** To use this YAML file for our use case, we need to

1. Go to the GitHub repository for the project
2. Create a folder named ***.github/workflows/***
3. In the above folder, create a file named ***pipeline.yml***
4. Copy and paste the above content for the YAML file into the ***pipeline.yml*** file

## Requirements file for the Github Actions Workflow

## Github Authentication and Push Files

* Before moving forward, we need to generate a secret token to push files directly from Colab to the GitHub repository.
* Please follow the below instructions to create the GitHub token:
    - Open your GitHub profile.
    - Click on ***Settings***.
    - Go to ***Developer Settings***.
    - Expand the ***Personal access tokens*** section and select ***Tokens (classic)***.
    - Click ***Generate new token***, then choose ***Generate new token (classic)***.
    - Add a note and select all required scopes.
    - Click ***Generate token***.
    - Copy the generated token and store it safely in a notepad.

In [ ]:
# Install Git
!apt-get install git

# Set your Git identity (replace with your details)
!git config --global user.email "<-------GitHub Email Address------->"
!git config --global user.name "<--------GitHub UserName--------->"

# Clone your GitHub repository
!git clone https://github.com/<--------GitHub UserName--------->/<--------GitHub Reponame--------->.git

# Move your folder to the repository directory
!mv /content/tourism_project/ /content/<--------GitHub Reponame--------->

In [ ]:
# Change directory to the cloned repository
%cd <--------GitHub Reponame--------->/

# Add the new folder to Git
!git add .

# Commit the changes
!git commit -m "first commit"

# Push to GitHub (you'll need your GitHub credentials; use a personal access token if 2FA enabled)
!git push https://<--------GitHub UserName--------->:<--------GitHub Token--------->@github.com/<--------GitHub UserName--------->/<--------GitHub Reponame--------->.git

# Output Evaluation

- GitHub (link to repository, screenshot of folder structure and executed workflow)

- Streamlit on Hugging Face (link to HF space, screenshot of Streamlit app)

<font size=6 color="navyblue">Power Ahead!</font>
___